### PR TITLE
feat: update react query to V5

### PR DESCRIPTION
**Pull request by MieszkoTSH**
Update react query to v5
Change use query abstraction to allow more granular cache management 
query factory was based on this article https://tkdodo.eu/blog/the-query-options-api. Since we use dependency injection to inject client I was not able to use react query built in `queryOptions` and `infiniteQueryOptions` util functions. 
Functions `queryFactoryOptions` `infiniteQueryFactoryOptions` are created for typesafety purposes only. They are more limited then the ones from the article (https://tkdodo.eu/blog/the-query-options-api#datatag - they won't infer proper types in e.g `queryClient.getQueryData`)

### CODE REVIEW

File: src/context/apiClient/apiClientContextController/apiError/apiError.ts [hunk 1]
```diff
@@ -1,31 +1,49 @@
...
 
-import { ApiError, BasicErrorData, FormErrorData } from './apiError.types';
+import { BasicApiError, BasicErrorData, FormApiError, FormErrorData, UnknownApiError } from './apiError.types';
 
-export const getStandardizedApiError = (error: AxiosError<unknown>): ApiError => {
+export class ApiError<T extends BasicApiError | FormApiError | UnknownApiError> extends Error {
+  originalError: T['originalError'];
+  statusCode: T['statusCode'];
+  type: T['type'];
+  data: T['data'];
```
Comment (tokerson) [inline src/context/apiClient/apiClientContextController/apiError/apiError.ts:10]:
```suggestion
  readonly originalError;
  readonly statusCode;
  readonly type;
  readonly data;
```
I think we can simplify it like that, cause types will be inferred based on the assignment in the constructor

File: src/context/auth/authContextController/AuthContextController.tsx [hunk 2]
```diff
@@ -37,19 +40,26 @@ export const AuthContextController = ({ children }: AuthContextControllerProps)
...
   } = useUser({
     enabled: !!state.accessToken,
-    onError: () => {
-      dispatch(resetTokens());
-    },
   });
 
+  const resetUser = useCallback(() => {
+    queryClient.removeQueries({ queryKey: authQueries.me().queryKey });
+  }, [queryClient]);
```
Comment (tokerson) [inline src/context/auth/authContextController/AuthContextController.tsx:50]: what do you think about moving this code back to the `useUser` hook? Thanks to that we will keep the previous contract of it and all the code regarding this query will be encapsulated there if needed to be used somewhere else

File: src/hooks/useQuery/useQuery.ts [hunk 3]
```diff
@@ -1,21 +1,23 @@
...
-  const queryKey: QueryKey = parseQueryKey(query, args);
+  const { queryFn, ...options } = params;
 
-  const result = useRQQuery(queryKey, async () => (await queryFn(args)) as DataForQuery<Key>, options);
+  const result = useRQQuery({
+    queryFn: (args) => queryFn(client)(args),
+    ...options,
+  });
 
   return { ...result, isLoadingAndEnabled: result.isLoading && result.fetchStatus !== 'idle' };
```
Comment (tokerson) [inline src/hooks/useQuery/useQuery.ts:22]:
```suggestion
  return { ...result, isLoadingAndEnabled: result.isPending && result.fetchStatus !== 'idle' };
```
Shouldn't it be changed like that cause `isLoading` was renamed to `isPending` and now `isLoading` is the previous `isInitialLoading`

File: src/routes/users/-components/UsersList.tsx [hunk 4]
```diff
@@ -3,20 +3,18 @@ import { useSearch, useNavigate } from '@tanstack/react-router';
...
-    'getUsersList',
-    { page: page.toString() || undefined },
-    {
-      select: (data) => {
-        return { ...data, users: data.users.sort((a, b) => (sort === 'desc' ? +b.id - +a.id : +a.id - +b.id)) };
-      },
+  const { data: usersResponse, isFetched: areUsersFetched } = useQuery({
+    ...authQueries.list({ page: page.toString() }),
+    select: (data) => {
+      return { ...data, users: data.users.sort((a, b) => (sort === 'desc' ? +b.id - +a.id : +a.id - +b.id)) };
```
Comment (tokerson) [inline src/routes/users/-components/UsersList.tsx:15]: I'm always surprised when I see this + - + : + - + :D looks a bit like https://en.wikipedia.org/wiki/Brainfuck

Comment (qh4r) [reply src/routes/users/-components/UsersList.tsx:15]: I think it's worth looking into whether we need the ability to sort here at all and if so, is sorting by id the way to go.

File: src/hooks/useQuery/useQuery.types.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+import { UseQueryOptions as UseRQQueryOptions } from '@tanstack/react-query';
+
+import { StandardizedApiError } from 'context/apiClient/apiClientContextController/apiError/apiError.types';
+
+export type UseQueryOptions<TData> = Omit<UseRQQueryOptions<TData, StandardizedApiError>, 'queryFn' | 'queryKey'>;
```
[the remark this answers is not in the document]
Comment (qh4r) [reply src/hooks/useQuery/useQuery.types.ts:5]: I think what we need here are 2 types sth like:

``` ts
// used exclusively in our custom useQuery hook
export type UseQueryOptions<TData, TError = StandardizedApiError> = Omit<
  UseRQQueryOptions<TData, TError>,
  'queryFn'
> & {
  meta?: Partial<ExtendedQueryMeta>;
  queryFn: (client: AxiosInstance) => QueryFunction<TData, QueryKey>;
};

// used with all of the dedicated query hooks in tandem with query factories
export type GenricQueryOptions<TData, TError = StandardizedApiError> = Omit<
  UseQueryOptions<TData, TError>,
  'queryKey' | 'queryFn'
>;
```

I'm not saying we should remove Axios client context and make it a global variable, or not let's discuss it, but I believe it's an entirely separate matter from this type above being easy to confuse with the build-in Tanstack one and hence misused.

[the remark this answers is not in the document]
Comment (tokerson) [reply src/hooks/useQuery/useQuery.types.ts:5]: I wonder if our comments are in the scope of this Pull Request. What do you think about deciding that it should not block merging and we could cover refactoring that part as a separate ticket as the goal of this one was to migrate to v5 of tantack-query?

File: src/context/apiClient/apiClientContextController/apiError/apiError.ts (hunk 1 above)
Comment (MieszkoTSH) [reply src/context/apiClient/apiClientContextController/apiError/apiError.ts:10]: Yeah it's a leftover as I had some issues with types when setting this up. I was figuring out whether it's possible to achieve proper type inference with just type `ApiError<BasicApiError | FormApiError | UnknownApiError>`

File: src/routes/users/-components/UsersList.tsx (hunk 4 above)
Comment (MieszkoTSH) [reply src/routes/users/-components/UsersList.tsx:15]: We might want to change / delete that in the future :D

File: src/api/actions/auth/auth.queries.ts [hunk 6]
```diff
@@ -1,29 +1,47 @@
...
-      return (await client.get<GetUsersResponse>(`/users/${queryParams}`)).data;
-    },
-  getUsersList:
-    (client: AxiosInstance) =>
-    async ({ page = '1' }: GetUsersListArgs) => {
-      const queryParams = stringify({ page, count: 5 }, { addQueryPrefix: true });
-      return (await client.get<GetUsersResponse>(`/users/${queryParams}`)).data;
-    },
-  // QUERY_FUNCTIONS_SETUP
+  all: () => ['users'],
```
Comment (tokerson) [inline src/api/actions/auth/auth.queries.ts:27]: I like how that was refactored really much 👍

**Review by tokerson — APPROVED** [review]

**Review by pzmuda1 — APPROVED** [review]

**Review by qh4r — APPROVED** [review]

**Pull request merged**